### PR TITLE
Evwork

### PR DESCRIPTION
**Pull request by kris-watts-gravwell**
EVs at ingest baby.

### CODE REVIEW

File: ingest/entry/entry.go [hunk 1]
```diff
@@ -45,108 +46,238 @@ type Entry struct {
...
 func (ent *Entry) Key() EntryKey {
 	return EntryKey(ent.TS.Sec)
 }
 
+// EnumeratedValues returns the slice of enumerated values, this is an accessor to prevent direct assignment
+func (ent Entry) EnumeratedValues() []EnumeratedValue {
+	return ent.evb.Values()
+}
+
+// ClearEnumeratedValues is a convienence function to remove all enumerated values
```
Comment (john-floren-gravwell) [inline ingest/entry/entry.go:61]: spelling

File: ingest/entry/entry.go [hunk 2]
```diff
@@ -45,108 +46,238 @@ type Entry struct {
...
+func (ent *Entry) AddEnumeratedValue(ev EnumeratedValue) (err error) {
+	if ev.Valid() {
+		ent.evb.Add(ev)
+	} else {
+		err = ErrInvalid
+	}
+	return
+}
+
+func (ent *Entry) AddEnumeratedValueEx(name string, val interface{}) error {
```
Comment (john-floren-gravwell) [inline ingest/entry/entry.go:75]: comment

File: ingest/entry/entry.go [hunk 3]
```diff
@@ -45,108 +46,238 @@ type Entry struct {
...
-func (ent *Entry) DecodeHeader(buff []byte) (int, error) {
+// decodeHeaderAlt gets a direct handle on the SRC buffer
+func (ent *Entry) decodeHeaderAlt(buff []byte) (int, bool) {
+	var hasEvs bool
+	var datasize uint32
+	ent.TS, ent.SRC, ent.Tag, hasEvs, datasize = DecodeHeader(buff)
+	return int(datasize), hasEvs
+}
+
+func (ent *Entry) DecodeHeader(buff []byte) (int, bool, error) {
```
Comment (john-floren-gravwell) [inline ingest/entry/entry.go:187]: comment: what are the return values?

File: ingest/entry/entry.go [hunk 4]
```diff
@@ -45,108 +46,238 @@ type Entry struct {
...
+	datasize &= flagMask // clear flags from datasize
+	hasEvs = ((flags & flagEVs) != 0)
+	ipv4 = ((flags & flagIPv4) != 0)
+	tag = EntryTag(binary.LittleEndian.Uint16(buff[16:]))
+	ts.Decode(buff[4:])
+	return
+}
+
+// decodeHeader copies copies the SRC buffer
+func (ent *Entry) decodeHeader(buff []byte) (int, bool) {
```
Comment (john-floren-gravwell) [inline ingest/entry/entry.go:170]: what are the return values?

File: ingest/entry/entry.go [hunk 5]
```diff
@@ -45,108 +46,238 @@ type Entry struct {
...
+	var datasize uint32
+	var src net.IP
+	ent.TS, src, ent.Tag, hasEvs, datasize = DecodeHeader(buff)
+	ent.SRC = append(net.IP(nil), src...)
+	return int(datasize), hasEvs
 }
 
-func (ent *Entry) DecodeHeader(buff []byte) (int, error) {
+// decodeHeaderAlt gets a direct handle on the SRC buffer
+func (ent *Entry) decodeHeaderAlt(buff []byte) (int, bool) {
```
Comment (john-floren-gravwell) [inline ingest/entry/entry.go:180]: ditto return values

File: ingest/entry/entry.go [hunk 6]
```diff
@@ -45,108 +46,238 @@ type Entry struct {
...
+		return 0, false, ErrInvalidBufferSize
 	}
-	return ent.decodeHeader(buff), nil
+	dataLen, hasEvs := ent.decodeHeader(buff)
+	return dataLen, hasEvs, nil
 }
 
 // DecodeEntry will copy values out of the buffer to generate an entry with its own
 // copies of data.  This ensures that entries don't maintain ties to blocks
 // DecodeEntry assumes that a size check has already happened
```
Comment (john-floren-gravwell) [inline ingest/entry/entry.go:197]: Remember that the docs ignores a single newline, so you need to end sentences with periods.

File: ingest/entry/entry.go [hunk 7]
```diff
@@ -45,108 +46,238 @@ type Entry struct {
...
+	if hasEvs {
+		var n int
+		if n, err = ent.evb.Decode(buff); err == nil {
+			off += n
+		}
+	}
+	return off, err
+}
+
+// DecodeAlt completely decodes an entry and returns the number of bytes consumed from a buffer
```
Comment (john-floren-gravwell) [inline ingest/entry/entry.go:247]: periods at the end of lines

File: ingest/entry/entry.go [hunk 8]
```diff
@@ -45,108 +46,238 @@ type Entry struct {
...
+		}
+		off += n
+	}
+	return off, nil
 }
 
-// EncodeHeader Encodes the header into the buffer for the file transport
-// think file indexer
-func (ent *Entry) EncodeHeader(buff []byte) error {
+// EncodeHeader Encodes the header into the buffer for the file transport think file indexer
```
Comment (john-floren-gravwell) [inline ingest/entry/entry.go:274]: fix up this comment for grammar and periods

**Review by john-floren-gravwell — CHANGES_REQUESTED** [review]

**Review by john-floren-gravwell — CHANGES_REQUESTED** [review]

File: ingest/api.go [hunk 9]
```diff
@@ -24,7 +24,7 @@ import (
 const (
 	//MAJOR API VERSIONS should always be compatible, there just may be additional features
 	API_VERSION_MAJOR uint32 = 0
-	API_VERSION_MINOR uint32 = 6
+	API_VERSION_MINOR uint32 = 8
```
[the remark this answers is not in the document]
Comment (kris-watts-gravwell) [reply ingest/api.go:27]: it is showing previous as 7 for me... but updating how this is handled

**Comment by kris-watts-gravwell** [comment]
PTAL

**Review by john-floren-gravwell — APPROVED** [review]

**Pull request merged**